### PR TITLE
Add ImageSharp project

### LINKED ISSUE #6: Create cross-plat package using ImageSharp
It would be great to have a cross-platform version of this library

### PR DESCRIPTION
**Pull request by slang25**
Fixes #6 

Needs tests!

### CODE REVIEW

File: Blurhash.ImageSharp/Encoder.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        public string Encode(Image<Rgba32> image, int componentsX, int componentsY)
+        {
+            return CoreEncode(ConvertBitmap(image), componentsX, componentsY);
+        }
+
+        /// <summary>
+        /// Converts the given bitmap to the library-independent representation used within the Blurhash-core
+        /// </summary>
+        /// <param name="sourceBitmap">The bitmap to encode</param>
+        internal static Pixel[,] ConvertBitmap<T>(Image<T> sourceBitmap) where T : struct, IPixel<T>
```
Comment (antonfirsov) [inline Blurhash.ImageSharp/Encoder.cs]: Considering the implementation,  you should consume `Image<Rgb24>` or use proper conversion with `PixelOperations<T>.ToRgb24(...)`.

File: Blurhash.ImageSharp/Encoder.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            
+            for (int y = 0; y < height; y++)
+            {
+                var rgbValues = MemoryMarshal.AsBytes(sourceBitmap.GetPixelRowSpan(y));
+
+                var index = stride;
+
+                for (var x = 0; x < width; x++)
+                {
+                    result[x, y].Red = MathUtils.SRgbToLinear(rgbValues[index]);
```
Comment (antonfirsov) [inline Blurhash.ImageSharp/Encoder.cs]: You will get incorrect results here with `Image<Rgba32>` or any other pixel type.

File: Blurhash.ImageSharp/Decoder.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        /// </summary>
+        /// <param name="pixelData">The library-independent representation of the image</param>
+        /// <returns>A <c>System.Drawing.Bitmap</c> in 32bpp-RGB representation</returns>
+        internal static Image<Rgb24> ConvertToBitmap(Blurhash.Core.Pixel[,] pixelData)
+        {
+            var width = pixelData.GetLength(0);
+            var height = pixelData.GetLength(1);
+
+            var data = Enumerable.Range(0, height)
+                .SelectMany(y => Enumerable.Range(0, width).Select(x => (x, y)))
```
Comment (antonfirsov) [inline Blurhash.ImageSharp/Decoder.cs]: Why not a for loop? This linq expression is a perf killer.

Comment (slang25) [reply Blurhash.ImageSharp/Decoder.cs]: It was copied from another implementation, I'll expand it, as I think perf is key if it's to be used in a Xamarin app for example, which is my primary use-case.

File: Blurhash.ImageSharp/Encoder.cs (hunk 2 above)
Comment (slang25) [reply Blurhash.ImageSharp/Encoder.cs]: I think I'll get the correct behaviour for `Rgba32` & `Rgb24`, basically and pixel where the first 24bits are RGB. The public methods guard for the supported types.
Have I understood your concern? Or have is there something that wont work?

Comment (antonfirsov) [reply Blurhash.ImageSharp/Encoder.cs]: > I think I'll get the correct behaviour for Rgba32 & Rgb24

Yeah that's actually true, I missed `bytesPerPixel`. However, you'll get the colors wrong with `Bgr24` or `Argb32` or anything else.

File: Blurhash.ImageSharp/Encoder.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            var width = sourceBitmap.Width;
+            var height = sourceBitmap.Height;
+            var bytesPerPixel = sourceBitmap.PixelType.BitsPerPixel / 8;
+            var stride = width * 3;
+
+            var result = new Pixel[width, height];
+            
+            for (int y = 0; y < height; y++)
+            {
+                var rgbValues = MemoryMarshal.AsBytes(sourceBitmap.GetPixelRowSpan(y));
```
Comment (antonfirsov) [inline Blurhash.ImageSharp/Encoder.cs]: `sourceBitmap.GetPixelRowSpan(y)` will be obsolete starting from RC1 (see  SixLabors/ImageSharp#1109). I suggest processing row-by-row, using `image.GetPixelRowSpan(y)`.

File: Blurhash.ImageSharp/Decoder.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                {
+                    for (int x = 0; x < width; x++)
+                    {
+                        var pixel = pixelData[x, y];
+                        yield return (byte) MathUtils.LinearTosRgb(pixel.Red);
+                        yield return (byte) MathUtils.LinearTosRgb(pixel.Green);
+                        yield return (byte) MathUtils.LinearTosRgb(pixel.Blue);
+                    }
+                }
+            }
```
Comment (antonfirsov) [inline Blurhash.ImageSharp/Decoder.cs]:
```suggestion
            Rgb24[] GenerateBitmap()
            {
                Rgb24[] result = new Rgb24[width*heitht];
                for (int y = 0; y < height; y++)
                {
                    Span<Rgb24> rowSpan = result.AsSpan().Slice(y*width, width);
                    for (int x = 0; x < width; x++)
                    {
                        var pixel = pixelData[x, y];
                        ref Rgb24 dest = ref rowSpan[x];
                        dest.R = (byte) MathUtils.LinearTosRgb(pixel.Red);
                        dest.G = (byte) MathUtils.LinearTosRgb(pixel.Green);
                        dest.B = (byte) MathUtils.LinearTosRgb(pixel.Blue);
                    }
                }
                return result;
            }
```
(Not tested)

`IEnumerable<AnyStruct>` is really slow when dealing with large amounts of data.

Comment (slang25) [reply Blurhash.ImageSharp/Decoder.cs]: Nice 👍, I'll give it a spin

File: Blurhash.ImageSharp/Encoder.cs (hunk 2 above)
Comment (slang25) [reply Blurhash.ImageSharp/Encoder.cs]: I've added a guard now to constrain the pixel type

File: Blurhash.ImageSharp/Decoder.cs (hunk 5 above)
Comment (slang25) [reply Blurhash.ImageSharp/Decoder.cs]: It works great 🙂

File: Blurhash.ImageSharp/Encoder.cs (hunk 2 above)
Comment (antonfirsov) [reply Blurhash.ImageSharp/Encoder.cs]: Missed that as well ^^

Comment (ithiredguns) [reply Blurhash.ImageSharp/Encoder.cs:64]: rgbValues[index] throws outofbounds error

Comment (JimBobSquarePants) [reply Blurhash.ImageSharp/Encoder.cs:64]: @slang25 Do you reckon it would be possible to give me write access to your fork? I have ideas to polish this.

Comment (slang25) [reply Blurhash.ImageSharp/Encoder.cs:64]: Done 😃

**Comment by Daniel15** [comment]
Will this be merged at some point? :)

**Comment by JimBobSquarePants** [comment]
I need to have a hack at this. I was given permission, just haven’t gotten to it yet.

**Comment by MarkusPalcer** [comment]
Sorry, I seeem to have notifications setup completely wrong - I didn't get notified that there was activity on the repo, so I didn't react to anything.

**Comment by MarkusPalcer** [comment]
> 
> 
> Has this been published to NuGet yet?
> 
> Will this be merged at some point? :)

To publish it on NuGet a .nuspec file is needed for the project.
I'm currently setting up GitHub actions, so all NuGet packages of the repo will be published in a feed

**Pull request merged**